File: src/commands/bwr.ts
import { SlashCommandBuilder, ChatInputCommandInteraction, EmbedBuilder } from "discord.js";
import { reply } from "../lib/functions/discord";
import { parseTime } from "../lib/functions/util";
import userSchema from "../lib/schemas/user";
import modeMap from "gokz.js/lib/api";
import SchnoseBot from "src/classes/Schnose";
import { bwr_wasm } from "../../rust/pkg/gokz_wasm";
import * as W from "src/lib/types/wasm";

export default {
	data: new SlashCommandBuilder()
		.setName("bwr")
		.setDescription("Check the World Record on a bonus.")
		.addStringOption((o) => o.setName("map").setDescription("Specify a map.").setRequired(true))
		.addIntegerOption((o) => o.setName("course").setDescription("Specify a bonus."))
		.addStringOption((o) =>
			o.setName("mode").setDescription("Specify a mode.").setChoices(
				{
					name: "KZT",
					value: "kz_timer"
				},
				{
					name: "SKZ",
					value: "kz_simple"
				},
				{
					name: "VNL",
					value: "kz_vanilla"
				}
			)
		),

	async execute(interaction: ChatInputCommandInteraction, client: SchnoseBot) {
		await interaction.deferReply();

		const inputMap = interaction.options.getString("map")!;
		const inputCourse = interaction.options.getInteger("course") || 0;
		const inputMode = interaction.options.getString("mode") || null;

		let mode: string;
		if (inputMode) mode = inputMode;
		else {
			const userDB = await userSchema.find({ discordID: interaction.user.id });
			if (!userDB[0]?.mode)
				return reply(interaction, {
					content: "You must either specify a mode or set a default value using `/mode`."
				});
			else mode = userDB[0].mode;
		}

		const request = await bwr_wasm(inputMap, inputCourse, mode);

		const result: W.bwr_wasm[] = [null, null];
		try {
			const temp = JSON.parse(request);
			for (let i = 0; i < temp.length; i++) {
				try {
					result[i] = JSON.parse(temp[i]) as W.bwr_wasm;
				} catch (_) {
					result[i] = null;
				}
			}
		} catch (_) {
			return reply(interaction, { content: request });
		}

		if (!result[0] && !result[1]) return reply(interaction, { content: request });

		const embed = new EmbedBuilder()
			.setColor([116, 128, 194])
			.setTitle(`[BWR ${inputCourse}] - ${result[0]?.map_name || result[1]?.map_name}`)
			.setURL(
				`https://kzgo.eu/maps/${result[0]?.map_name || result[1]?.map_name}&bonus=${inputCourse}`
			)
			.setThumbnail(
				`https://raw.githubusercontent.com/KZGlobalTeam/map-images/master/images/${
					result[0]?.map_name || result[1]?.map_name
				}.jpg`
			)
			.addFields([
				{
					name: "TP",
					value: `${result[0]?.time ? parseTime(result[0].time) : "😔"} ${
						result[0]?.player_name ? `(${result[0].player_name})` : ""
					}`,
					inline: true
				},
				{
					name: "PRO",
					value: `${result[1]?.time ? parseTime(result[1].time) : "😔"} ${
						result[1]?.player_name ? `(${result[1].player_name})` : ""
					}`,
					inline: true
				}
			])
			.setFooter({
				text: `(͡ ͡° ͜ つ ͡͡°)7 | Mode: ${modeMap.get(result[0]?.mode || result[1]?.mode)}`,
				iconURL: client.icon
			});

		return reply(interaction, {
			embeds: [embed]
		});
	}
};

File: src/commands/bwr.rs
use {
	crate::{
		prelude::InteractionResult,
		events::interactions::InteractionState,
		database::util as DB,
		formatting::{get_replay_links, format_time, attach_replay_links},
	},
	gokz_rs::{
		prelude::*,
		global_api::{get_maps, is_global, get_wr},
	},
	serenity::{
		builder::{CreateApplicationCommand, CreateEmbed},
		model::prelude::command::CommandOptionType,
	},
	futures::future::join_all,
	itertools::Itertools,
};

pub(crate) fn register(cmd: &mut CreateApplicationCommand) -> &mut CreateApplicationCommand {
	return cmd
		.name("bwr")
		.description("Check the World Record on a bonus.")
		.create_option(|opt| {
			opt.kind(CommandOptionType::String)
				.name("map_name")
				.description("Specify a map.")
				.required(true)
		})
		.create_option(|opt| {
			opt.kind(CommandOptionType::String)
				.name("mode")
				.description("Choose a mode.")
				.add_string_choice("KZT", "kz_timer")
				.add_string_choice("SKZ", "kz_simple")
				.add_string_choice("VNL", "kz_vanilla")
				.required(false)
		})
		.create_option(|opt| {
			opt.kind(CommandOptionType::Integer)
				.name("course")
				.description("Specify a course.")
				.required(false)
		});
}

pub(crate) async fn execute(state: &mut InteractionState<'_>) -> InteractionResult {
	// Defer current interaction since this could take a while
	state.defer().await?;

	let map_name = state.get::<String>("map_name").expect("This option is marked as `required`.");
	let course = state.get::<u8>("course").unwrap_or(1);

	let mode = match state.get::<String>("mode") {
		Some(mode_name) => mode_name
			.parse()
			.expect("The possible values for this are hard-coded and should never be invalid."),
		None => DB::fetch_mode(state.user, state.db, true).await?,
	};

	let global_maps = get_maps(state.req_client).await?;

	let map = is_global(&MapIdentifier::Name(map_name), &global_maps).await?;

	let map_identifier = MapIdentifier::Name(map.name.clone());

	let (tp, pro) = join_all([
		get_wr(&map_identifier, &mode, true, course, state.req_client),
		get_wr(&map_identifier, &mode, false, course, state.req_client),
	])
	.await
	.into_iter()
	.collect_tuple()
	.expect("This cannot fail, look 6 lines up.");

	if let (&Err(_), &Err(_)) = (&tp, &pro) {
		return Ok("No BWRs found 😔.".into());
	}

	let (player_tp, player_pro) = {
		let (mut player_tp, mut player_pro) = (String::from("unknown"), String::from("unknown"));
		if let Ok(rec) = &tp {
			if let Some(name) = rec.player_name.clone() {
				player_tp = name;
			}
		}
		if let Ok(rec) = &pro {
			if let Some(name) = rec.player_name.clone() {
				player_pro = name;
			}
		}
		(player_tp, player_pro)
	};

	let links = (get_replay_links(&tp).await, get_replay_links(&pro).await);

	let mut embed = CreateEmbed::default()
		.colour(state.colour)
		.title(format!("[BWR {}] {} (T{})", &course, &map.name, &map.difficulty))
		.url(format!(
			"{}?bonus={}&{}=",
			state.map_link(&map.name),
			&course,
			&mode.to_fancy().to_lowercase()
		))
		.thumbnail(state.map_thumbnail(&map.name))
		.field(
			"TP",
			format!(
				"{} ({})",
				match &tp {
					Ok(rec) => format_time(rec.time),
					_ => String::from("😔"),
				},
				player_tp
			),
			true,
		)
		.field(
			"PRO",
			format!(
				"{} ({})",
				match &pro {
					Ok(rec) => format_time(rec.time),
					_ => String::from("😔"),
				},
				player_pro
			),
			true,
		)
		.footer(|f| f.text(format!("Mode: {}", mode.to_fancy())).icon_url(state.icon))
		.to_owned();

	attach_replay_links(&mut embed, links);

	Ok(embed.into())
}
